### DISCUSSION TITLE
Handling Salesforce paginated apis with Quarkus

### DISCUSSION THREAD

**Discussion by IvanPuntev** [quarkusio/quarkus#36681] [Q&A] [answered]
Hi,

Me and my team recently started using quarkus and we have a case where we need to handle the Salesforce api pagination.
For example, the returned json response from the api is like this:

{
  "totalSize": 3222,
  **"done": false,**
  **"nextRecordsUrl": "/services/data/v59.0/query/01gRO0000016PIAYA2-500",**
  "records": [
    {
      "attributes": {
        "type": "Contact",
        "url": "/services/data/v59.0/sobjects/Contact/003RO0000035WQgYAM"
      },
      "Id": "003RO0000035WQgYAM",
      "Name": "John Smith"
    },
    ...
  ]
}

Instead of a page number we have a **nextRecordsUrl** field containing the id for the next batch of records. The first response contains the id to the second batch, the second response contains the id to the third, etc until the field **done** is equal to true.
The only pagination I could find is this - https://quarkus.io/blog/mutiny-pagination/

Can you help us with such example using mutiny for our case?

**Comment by alex-martel** [comment, marked as answer]
Something like this

```java
     Multi<SalesforceRecord> multi = Multi.createBy().repeating()
                .uni(() -> new AtomicReference<String>("firstPage"),
                        url -> sfClient.getRecord(url.get()).onItem()
                                .invoke(sfRecord -> url.set(sfRecord.getNextRecordsUrl())))
                .until(SalesforceRecord::isDone);
```

**Reply by IvanPuntev** [reply]
Thanks!

**Reply by IvanPuntev** [reply]
@manofthepeace When I try to compile the code with maven I get this error for the  part with .until(SalesforceRecord::isDone);

[ERROR]     incompatible types: invalid method reference
[ERROR]     method isDone in class ......SalesforceRecord cannot be applied to given types
[ERROR]       required: no arguments
[ERROR]       found:    java.lang.Object
[ERROR]       reason: actual and formal argument lists differ in length

For some reason it doesn't like the method reference. It is compiled successfully if lambda is used - .until(sfRecord -> sfRecord.isDone());

**Reply by IvanPuntev** [reply]
until() does not propagate the item that did pass the check so I had to replace it with whilst(sfRecord -> !sfRecord.isDone()) which propagates the item downstream even if the check does not pass.
